File: content/snippets/scala2/interview_lcci_16_15_master_mind.scala
object Lcci1615MasterMind {
    def masterMind(solution: String, guess: String): Array[Int] = {
        var hits = 0
        val cs = Array.fill(26)(0)
        val cg = Array.fill(26)(0)
        for (i <- 0 until 4) {
            if (solution(i) == guess(i)) {
                hits += 1
            } else {
                cs(solution(i) - 'A') += 1
                cg(guess(i) - 'A') += 1
            }
        }
        var pseudo = 0
        for (i <- 0 until 26) {
            pseudo += math.min(cs(i), cg(i))
        }
        Array(hits, pseudo)
    }

    def main(args: Array[String]): Unit = {
        println(masterMind("RGBY", "GGRR").mkString("[", ", ", "]"))
    }
}

File: content/snippets/typescript/interview_lcci_16_15_master_mind.ts
function masterMind(solution: string, guess: string): number[] {
    let hits = 0;
    const cs = new Array<number>(26).fill(0);
    const cg = new Array<number>(26).fill(0);
    for (let i = 0; i < 4; i++) {
        if (solution[i] === guess[i]) {
            hits += 1;
        } else {
            cs[solution.charCodeAt(i) - 65] += 1;
            cg[guess.charCodeAt(i) - 65] += 1;
        }
    }
    let pseudo = 0;
    for (let i = 0; i < 26; i++) {
        pseudo += Math.min(cs[i], cg[i]);
    }
    return [hits, pseudo];
}

console.log(masterMind("RGBY", "GGRR"));
